Sumo-Logic-Tools/Event_Based_S3_Automation/script.py

```diff
@@ -0,0 +1,265 @@
+
+		while True:
+			deployment = int(input("Enter your choice: "))
+			if (deployment < 1 or deployment > 5):
+				print("Please select a value between 1 to 5")
+				continue
+			else:
+				break
+
+		base_url = str(self.apiEndpointOrderedList[deployment - 1])
+
+		print("\nEnter you SUMO ACCESS ID AND ACCESS KEY")
+
+		access_id = input("ACCESS ID: ")
+		access_key = input("ACCESS KEY: ")
+
+		return (base_url, access_id, access_key)
+
+
+'''
+Helps in getting the list of sources on which event based S3 has to be set up.
+It does the following tasks:
+1. Get a list of all the collectors.
+2. Get a list of all the sources in every collector.
+3. Extract the list of polling sources on which event-based-S3 can be set up.
+4. Create a map between bucket name and endpoint and return it for further processing. So if there are multiple sources collecting from the same bucket 
+then the map can look like: Map("bucket1" -> ["endpoint1", "endpoint2"], "bucket2" -> ["endpoint3"]).
+'''
+class SumoSetupHelper():
+
+	collector_url = '/v1/collectors/'
+	pollingTypeSources = ['AwsS3Bucket', 'AwsCloudTrailBucket', 'AwsCloudFrontBucket', 'AwsConfigBucket', 'AwsElbBucket', 'AwsS3AuditBucket']
+	mapOfBucketNameToEndpoint = defaultdict(list)
+
+	def __init__(self, base_url, access_id, access_key):
+		self.base_url = base_url
+		self.access_id = access_id
+		self.access_key = access_key
+
+	def calculateMappingOfBucketNameToEndpoint(self):
+		listOfCollectors = self.getListOfCollectors()
+
+		listOfSourcesLink = self.getSourcesLinks(listOfCollectors)
+
+		listOfPollingSources = self.getPollingSources(listOfSourcesLink)
+
+		self.populateMapOfBucketNameToEndpoint(listOfPollingSources)
+
+		return self.mapOfBucketNameToEndpoint
+
+	def getListOfCollectors(self):
+		return self.makeGetRequest(self.collector_url).json()
+
+	def makeGetRequest(self, url):
+		return requests.get(url = self.base_url + str(url), auth = HTTPBasicAuth(self.access_id, self.access_key))
+
+	def getSourcesLinks(self, listOfCollectors):
+		listOfSourcesLink = list()
+		try:
+			for collector in listOfCollectors['collectors']:
+				links = collector['links']
+				listOfSourcesLink.extend(self.extractSourcesLink(collector['links']))
+		except KeyError as e:
+			try:
+				collector = listOfCollectors['collector']
+				links = collector['links']
+				listOfSourcesLink.extend(self.extractSourcesLink(collector['links']))
+			except KeyError as e:
+				print("The API did not return any collectors!")
+		return listOfSourcesLink
+
+	def extractSourcesLink(self, links):
+		listOfSourcesLink = list()
+		for link in links:
+			if (link['rel'] == 'sources'):
+				listOfSourcesLink.append(link['href'])
+
+		return listOfSourcesLink
+
+	def getPollingSources(self, listOfSourcesLink):
+		listOfPollingSources = list()
+		for sourceLink in listOfSourcesLink:
+			allSources = self.makeGetRequest(sourceLink).json()
```

Comment:
from collector management API we have found these types of calls in a tight loop are prone to throttling (we limit API requests to 4 per second, or every 250ms). For this reason we added a sleep in between successive API calls: https://github.com/SumoLogic/collector-management-client/blob/master/sumo_mgmt.py#L278